### ISSUE TITLE
CodeLens for non-function values

### ISSUE THREAD

**Issue by isaacabraham** [ionide/ionide-vscode-fsharp#178]
New codelens support (2.6) for functions is awesome. Having this also for non-function values would be great as well though. Perhaps having this optional via some setting would be needed though as it might not be for everyone.

**Comment by Krzysztof-Cieslak** [comment]
There are some problems on FSAC / FCS side of things with it (current method of getting symbols in file used by CodeLens and `Ctrl + Shfit + O` doesn't return value bindings. ) which makes adding this more complex than expected. It's something I'd want, but not super high priority.

**Comment by rojepp** [comment]
This also applies to vals that are functions. 
<img width="450" alt="screen shot 2016-10-04 at 22 35 00" src="https://cloud.githubusercontent.com/assets/445888/19091439/e53521a6-8a82-11e6-9948-3e565d78c1cc.png">

**Comment by rkosafo** [comment]
Sweet feature. Issue also applies to `pattern matching function`

![codelens](https://cloud.githubusercontent.com/assets/21700034/19096013/d9a18982-8a88-11e6-8db2-82d8604f4b5c.png)

**Comment by Krzysztof-Cieslak** [comment]
Worth noticing that if value is uppercase it will work.

**Comment by Krzysztof-Cieslak** [comment]
I think it's actually FCS bug, probably somewhere in https://github.com/fsharp/FSharp.Compiler.Service/blob/master/src/fsharp/vs/ServiceNavigation.fs#L82-L259

**Comment by 7sharp9** [comment]
afaik `ServiceNavigation` was never put into production so its a just one of its limitations.  The good news is you can get all the info you want from the symbol api.  If theres bugs in that then its my fault as I added it some of it.  :-)

**Comment by vasily-kirichenko** [comment]
I think this function misses `SynModuleDecl.Let` case: https://github.com/fsharp/FSharp.Compiler.Service/blob/8ecb153f1a0a9809e6283cd4531f4322c63ebd08/src/fsharp/vs/ServiceNavigation.fs#L214 

We collect navigable symbols ourselves in VFPT, the relevant function is https://github.com/fsprojects/VisualFSharpPowerTools/blob/master/src/FSharp.Editing/Navigation/NavigableItemsCollector.fs#L192, specifically https://github.com/fsprojects/VisualFSharpPowerTools/blob/master/src/FSharp.Editing/Navigation/NavigableItemsCollector.fs#L198

If nobody is working on fixing this bug, I can try to do it later today.

**Comment by vasily-kirichenko** [comment]
OK, I've added `SynPat.Named` to navigable items here https://github.com/vasily-kirichenko/FSharp.Compiler.Service/blob/add-module-values-to-navigable-items/src/fsharp/vs/ServiceNavigation.fs#L136-L137

Now for 

```fsharp
module Test1

let foo x = x

let intVal = 1

let stringVal = "foo"

[<EntryPoint>]
let main argv =
    printfn "%A" argv
    0
```

`declarations` returns the following:

```json
[
    {
        "Declaration": {
            "UniqueName": "Test1_1_of_1",
            "Name": "Test1",
            "Glyph": "Module",
            "GlyphChar": "N",
            "IsTopLevel": true,
            "Range": {
                "StartColumn": 1,
                "StartLine": 1,
                "EndColumn": 6,
                "EndLine": 12
            },
            "BodyRange": {
                "StartColumn": 13,
                "StartLine": 1,
                "EndColumn": 6,
                "EndLine": 12
            },
            "File": "e:\\github\\Test1\\Test1\\Test1.fs"
        },
        "Nested": [
            {
                "UniqueName": "Test1_1_of_1",
                "Name": "foo",
                "Glyph": "Field",
                "GlyphChar": "F",
                "IsTopLevel": false,
                "Range": {
                    "StartColumn": 5,
                    "StartLine": 3,
                    "EndColumn": 14,
                    "EndLine": 3
                },
                "BodyRange": {
                    "StartColumn": 5,
                    "StartLine": 3,
                    "EndColumn": 14,
                    "EndLine": 3
                },
                "File": "e:\\github\\Test1\\Test1\\Test1.fs"
            },
            {
                "UniqueName": "Test1_1_of_1",
                "Name": "intVal",
                "Glyph": "Field",
                "GlyphChar": "F",
                "IsTopLevel": false,
                "Range": {
                    "StartColumn": 5,
                    "StartLine": 5,
                    "EndColumn": 11,
                    "EndLine": 5
                },
                "BodyRange": {
                    "StartColumn": 5,
                    "StartLine": 5,
                    "EndColumn": 11,
                    "EndLine": 5
                },
                "File": "e:\\github\\Test1\\Test1\\Test1.fs"
            },
            {
                "UniqueName": "Test1_1_of_1",
                "Name": "main",
                "Glyph": "Field",
                "GlyphChar": "F",
                "IsTopLevel": false,
                "Range": {
                    "StartColumn": 5,
                    "StartLine": 10,
                    "EndColumn": 6,
                    "EndLine": 12
                },
                "BodyRange": {
                    "StartColumn": 5,
                    "StartLine": 10,
                    "EndColumn": 6,
                    "EndLine": 12
                },
                "File": "e:\\github\\Test1\\Test1\\Test1.fs"
            },
            {
                "UniqueName": "Test1_1_of_1",
                "Name": "stringVal",
                "Glyph": "Field",
                "GlyphChar": "F",
                "IsTopLevel": false,
                "Range": {
                    "StartColumn": 5,
                    "StartLine": 7,
                    "EndColumn": 14,
                    "EndLine": 7
                },
                "BodyRange": {
                    "StartColumn": 5,
                    "StartLine": 7,
                    "EndColumn": 14,
                    "EndLine": 7
                },
                "File": "e:\\github\\Test1\\Test1\\Test1.fs"
            }
        ]
    }
]
```
result:

![image](https://cloud.githubusercontent.com/assets/873919/19343698/f1a09c72-913f-11e6-9212-f1d3cc4bb1ff.png)

I think there are no code lens because `F` nested declarations are filtered out here https://github.com/ionide/ionide-vscode-fsharp/blob/master/src/Components/CodeLens.fs#L25

I cannot run experimental Code instance, so I cannot fix / test the ionide part.

**Comment by vasily-kirichenko** [comment]
OK, now it looks like this:

![image](https://cloud.githubusercontent.com/assets/873919/19345271/4f68aa4c-9146-11e6-91fc-a04d678de640.png)

So we have class's internal `let` binding code lensed, too. Is it OK?

**Comment by sergey-tihon** [comment]
Why not? At least I like it!

**Comment by vasily-kirichenko** [comment]
:) @sergey-tihon what do you think about not showing code lens on interface members as it duplicate signature 1:1?

**Comment by sergey-tihon** [comment]
Interesting idea... but i think that it probably should work on virtual methods...
```fsharp
type IPrintable =
   abstract Print : unit -> unit // do not show here
   default __.Print () = () // but show here
```

**Comment by vasily-kirichenko** [comment]
Current status:

![image](https://cloud.githubusercontent.com/assets/873919/19349607/93ca3b36-915c-11e6-81f0-88616c8d5df6.png)

I've no idea why it does not show signatures for default method implementations.

**Comment by vasily-kirichenko** [comment]
aha, it seems to relate to tooltip parsing on Code side:

![image](https://cloud.githubusercontent.com/assets/873919/19349710/2930109c-915d-11e6-8fa2-1fb4e3982699.png)

**Comment by vasily-kirichenko** [comment]
Implemented via https://github.com/ionide/ionide-vscode-fsharp/pull/203

**Comment by Krzysztof-Cieslak** [comment]
Released in `2.6.10`
